linalg/sparsemat.cpp

```diff
@@ -580,23 +609,78 @@ void SparseMatrix::AddMult(const Vector &x, Vector &y, const double a) const
    }
 
 #ifndef MFEM_USE_LEGACY_OPENMP
+
    const int height = this->height;
    const int nnz = J.Capacity();
    auto d_I = Read(I, height+1);
    auto d_J = Read(J, nnz);
    auto d_A = Read(A, nnz);
    auto d_x = x.Read();
    auto d_y = y.ReadWrite();
-   MFEM_FORALL(i, height,
+
+   //Skip if matrix is all zero
+   if (nnz == 0) {return;}
+#ifdef MFEM_USE_CUDA
+   if (Device::Allows(Backend::CUDA_MASK))
```

Comment:
Why is this the right condition for deciding between cuSPARSE and the native version? (This is a question, not a critique.) Also for my understanding, this decision is made at compile time, right? If so, we need to be sure that cuSPARSE is always faster and does not use significantly more GPU memory.

Reply:
From my small benchmarks I have found that CuSparse is much faster than the MFEM implementation. 
 Unfortunately, I think it may be sparse matrix dependent. For example 9 using PR https://github.com/mfem/mfem/pull/1506 cusparse provided a good speed up. But it could be that there exist a family of matrices where the native implementation can compete or out perform cusparse (if so how much?) 

Figure 2 of this paper: https://people.mpi-inf.mpg.de/~rzayer/pappe/HowNaiveIsNaiveSpmvOnTheGpu.pdf shows there some of matrices where ``naive`` version does out perform cusparse

Reply:
Thanks for clarifying, @artv3. Considering that there may be some cases where the native version is faster, and that cuSPARSE may use more memory, should we provide the user with a build switch to choose between the two? As it is now, it looks like cuSPARSE will always be used when MFEM is built with CUDA.

Reply:
I think that is reasonable, would a compile time switch be good? Something like MFEM_USE_CUSPARSE ?

Reply:
Yes, that is what I had in mind. It is fine to have it set to true by default, as it seems generally better than the native version.

Reply:
It seems that a runtime option in the SparseMatrix class will be more flexible and probably easier to implement (considering that we have 2 separate build systems to update if we add `MFEM_USE_CUSPARSE`).

Reply:
That sounds good to me, I can do that.

Reply:
Done! I added an additional method to the SparseMatrix class UseCuSparse(bool usecuSparse = true).

Reply:
I think it would be better to have `MFEM_USE_CUDA` implying `MFEM_USE_CUSPARSE`, like this someone that doesn't want to use `CuSparse` for some reason could still do `MFEM_USE_CUDA=YES MFEM_USE_CUSPARSE=NO`, but the default would be to use `CuSparse` with `Cuda`.

Reply:
@YohannDudouit, I wanted to make the choice of `cuSparse` matvec vs our simple matvec a runtime option (default is `cuSparse`) -- there is no need to introduce a compile time option for something that can easily be a runtime option. I see that you didn't like that for some reason. Can you explain why? Everyone else agreed with that suggestion.

Reply:
I was having doubts about the granularity of the method, local to each sparse matrix vs global to all of them. I find strange that it's a method, because it means it could change over time. I also don't really like methods that are about a property that does not always exist, e.g. when computing on AMD GPUs what does `A.UseCuSparse(true)` means?
What happens if we want to integrate another sparse matrix library, do we also add `UseHipSparse`? I think it would make more sense for this property to be part of the constructor and take a mask like you often do to specify non-default behavior. Are we sure we don't want a way to modify globally the defaults?

Reply:
That is a good point, we can have a ``UseVendorSparseMatVec`` option if MFEM supports other vendor solutions. Intel and AMD accelerators are indeed coming.  Perhaps there might be a way to set it globally once - at run time for all sparse matrices?

Reply:
The point about the name is very good, it may be best to use name like `SetMatvecAlgorithm` (or more generally `SetBackend` if the backend can be used for other operations too) with a parameter that comes from a new enumeration. The default value should probably depend on the currently configured `mfem::Device`.

Regarding making this option global: I agree that having a global version of `SetMatvecAlgorithm`/`SetBackend` will be convenient -- this global value can become the default value for new `SparseMatrix` objects (if we want to keep the object-local option too) or become a global switch for all `SparseMatrix` objects (if we remove the object-local option).

Reply:
Let's keep this for now, we'll change the `UseCuSparse()` method in the future if we need.